File: frontend/src/types/contracts.ts
/**
 * Canonical Data Contracts: AI In-Car Concierge ("Suzanne")
 * Universal architecture: EV-focused, agnostic routing, voice-first HMI.
 */

export type RoutePriority = 'FASTEST' | 'CHEAPEST' | 'SCENIC' | 'BALANCED';
export type RouteSessionStatus =
  | 'ROUTE_READY'
  | 'CHARGING_OPTIONS_READY'
  | 'CHARGING_CONFIRMED'
  | 'REROUTING'
  | 'DRIVING'
  | 'ERROR';

export type StopCategory =
  | 'charging'
  | 'food'
  | 'rest'
  | 'toll'
  | 'vignette'
  | 'service'
  | 'hotel'
  | 'restaurant'
  | 'attraction'
  | 'coffee'
  | 'toilets'
  | 'fuel'
  | 'shopping';

export interface VehicleState {
  vehicleId: string;
  propulsion: 'BEV';
  batteryPercent: number;          // 0 - 100
  estimatedRangeKm: number;
  consumptionRateKwh: number;      // kWh per 100km
  tyres: 'SUMMER' | 'WINTER' | 'ALL_SEASON';
  odometerKm: number;
}

export interface Coordinates {
  lng: number;
  lat: number;
}

export interface StopPinpoint {
  id: string;
  name: string;
  category: StopCategory;
  coords: [number, number];        // [lng, lat]
  rating?: number;
  userReviewCount?: number;
  tag: string;                     // e.g., "Fast Charger · 250kW" or "Italian Dining"
  details?: string | null;
  photoReference?: string | null;
  keywords?: string[];
  provider?: string | null;
  source?: string | null;
  amenities?: string[];
  distanceMeters?: number;
  detourMinutes: number;
  routeOffsetKm?: number;
  estimatedDrivingDetourMinutes?: number;
  chargingDurationMinutes?: number;
  connectorTypes?: string[];
  availability?: boolean | null;
  mandatory?: boolean;
  partner?: {
    id: string;
    name: string;
    benefit?: string | null;
    benefitScope?: string | null;
    benefitSource?: 'fixture' | 'configured' | 'provider' | null;
    verified?: boolean;
    status?: 'suggested' | 'confirmed' | 'completed';
  } | null;
  partnerBenefit?: string | null;
}

export interface PartnerFact {
  partnerId: string;
  brand: string;
  benefit?: string | null;
  benefitScope?: string | null;
  benefitSource: 'fixture' | 'configured' | 'provider';
  verified: boolean;
}

export interface TelemetryNarrationFacts {
  batteryPercent: number;
  estimatedRangeKm: number;
  maxChargedRangeKm: number;
  consumptionRateKwh: number;
  chargingFeasible?: boolean | null;
}

export interface RouteOpportunity {
  id: string;
  type: 'charging' | 'hotel' | 'restaurant' | 'amenity' | 'partner';
  stopId?: string | null;
  resultId?: string | null;
  partnerFact?: PartnerFact | null;
  reason: string;
  detourMinutes: number;
  requiresRouteConfirmation: boolean;
  status: 'suggested' | 'confirmed' | 'completed';
}

export interface ChargingPlan {
  stops: StopPinpoint[];
  complete: boolean;
  totalChargingMinutes: number;
  confirmed: boolean;
}

export interface ChargingOption {
  optionNumber: number;
  stop: StopPinpoint;
  status: 'suggested' | 'selected' | 'validated' | 'confirmed';
}

export interface RouteSessionFacts {
  chargingPlanConfirmed: boolean;
  confirmedChargingStopIds: string[];
  purchasedVignetteRequirementIds: string[];
  remainingRequirements: RouteRequirement[];
}

export interface RouteAlert {
  type: 'WEATHER' | 'TRAFFIC' | 'TOLL' | 'VEHICLE';
  locationName?: string;
  severity: 'INFO' | 'WARNING' | 'CRITICAL';
  message: string;                 // Concise driver text
}

export interface TripStats {
  totalDistanceKm: number;
  drivingDurationMinutes: number;
  totalDurationMinutes: number;
  totalPriceEur: number;           // Aggregated estimates (tolls + charging)
}

export interface BorderCrossing {
  fromCountry: string;
  toCountry: string;
}

export interface RouteRequirement {
  id: string;
  name: string;
  country: string;
  kind: 'toll' | 'vignette';
  mandatory: boolean;
}

export interface RouteResponse {
  sessionId?: string | null;
  routeStatus?: RouteSessionStatus;
  origin: string;
  originCoordinates?: Coordinates | null;
  destination: string;
  countries: string[];
  stats: TripStats;
  geometry: [number, number][];    // Full route polyline coordinates [[lng, lat], ...]
  stops: StopPinpoint[];
  alerts: RouteAlert[];
  chargingStop?: StopPinpoint | null;
  chargingRequired?: boolean;
  borderCrossings: BorderCrossing[];
  routeRequirements: RouteRequirement[];
  opportunities?: RouteOpportunity[];
  chargingOptions: ChargingOption[];
  chargingPlan?: ChargingPlan | null;
  sessionFacts?: RouteSessionFacts | null;
  telemetry?: TelemetryNarrationFacts | null;
}

export interface RoutePoiResponse {
  results: StopPinpoint[];
  opportunities?: RouteOpportunity[];
  routeId?: string | null;
  searchId?: string | null;
}

export interface StopAmenitiesResponse {
  selectedStopName: string;
  results: StopPinpoint[];
  opportunities?: RouteOpportunity[];
  radiusMeters: number;
  routeId: string;
  searchId?: string | null;
}

export interface ConfirmChargingStopResponse extends StopAmenitiesResponse {
  route: RouteResponse;
  chargingPlan?: ChargingPlan | null;
  sessionFacts?: RouteSessionFacts | null;
}

export interface PurchaseVignetteRequest {
  routeId: string;
  requirementId: string;
  confirmation: 'confirmed';
}

export interface PurchaseVignetteResponse {
  status: 'completed' | 'duplicate';
  sessionId?: string | null;
  transactionId: string;
  routeId: string;
  requirementId: string;
  walletStatus: 'ready' | 'processing' | 'completed' | 'declined' | 'duplicate';
  phoneConfirmationStatus: 'pending' | 'sent' | 'failed';
  amountEur: number;
  currency: 'EUR';
  sessionFacts?: RouteSessionFacts | null;
}

export type BookingType = 'hotel_room' | 'restaurant_table';

export interface BookingRequest {
  routeId: string;
  searchId: string;
  resultId: string;
  bookingType: BookingType;
  guests: number;
  date: string;
  time?: string | null;
  confirmation: 'confirmed';
}

export interface BookingResponse {
  status: 'completed' | 'duplicate' | 'pending' | 'failed';
  sessionId?: string | null;
  bookingId: string;
  resultId: string;
  routeId: string;
  bookingType: BookingType;
  guests: number;
  date: string;
  time?: string | null;
  walletStatus: 'ready' | 'processing' | 'completed' | 'declined' | 'duplicate';
  phoneConfirmationStatus: 'pending' | 'sent' | 'failed';
}

export interface StartDrivingResponse {
  status: 'active';
  sessionId?: string | null;
  routeId: string;
  remainingDistanceKm: number;
  remainingDurationMinutes: number;
  eta: string;
  nextStop?: Pick<StopPinpoint, 'id' | 'name' | 'category'> | null;
  chargingRequired: boolean;
}

export interface ReturnToMainRouteResponse {
  status: 'success';
  sessionId?: string | null;
  routeId: string;
}

export interface AssistantIntent {
  destination: string;
  priority: RoutePriority;
}

export interface AssistantResponse {
  transcript: string;
  intent: AssistantIntent;
  spokenResponse: string;          // Plain text fallback
  audioBase64?: string;            // MP3 audio bytes from OpenAI TTS
  route?: RouteResponse;           // Populated on Day 2; null on Day 1
  poiResults?: StopPinpoint[];
  toastMessage?: string;           // e.g. "ROUTE: BUDAPEST (FASTEST)"
}

export interface VehicleTelemetry {
  vehicleId: string;
  propulsion: 'BEV';
  batteryPercent: number;
  estimatedRangeKm: number;
  maxChargedRangeKm: number;
  consumptionRateKwh: number;
  connectorTypes: string[];
  maxChargingPowerKw: number;
  batteryCapacityKwh: number;
  tyres: 'SUMMER' | 'WINTER' | 'ALL_SEASON';
  odometerKm: number;
}

File: backend/app/models/contracts.py
from enum import StrEnum
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


def to_camel(value: str) -> str:
    first, *rest = value.split("_")
    return first + "".join(part.capitalize() for part in rest)


class ContractModel(BaseModel):
    model_config = ConfigDict(
        alias_generator=to_camel,
        populate_by_name=True,
        extra="forbid",
    )


class RoutePriority(StrEnum):
    FASTEST = "FASTEST"
    CHEAPEST = "CHEAPEST"
    SCENIC = "SCENIC"
    BALANCED = "BALANCED"


class AssistantIntent(ContractModel):
    destination: str = Field(min_length=1)
    priority: RoutePriority


class Coordinates(ContractModel):
    lng: float
    lat: float


class StopPinpoint(ContractModel):
    id: str
    name: str
    category: Literal[
        "charging", "hotel", "restaurant", "attraction", "coffee", "food",
        "rest", "toilets", "fuel", "toll", "vignette", "service", "shopping"
    ]
    coords: tuple[float, float]
    rating: float | None = None
    user_review_count: int | None = Field(default=None, ge=0)
    tag: str = ""
    details: str | None = None
    photo_reference: str | None = None
    keywords: tuple[str, ...] = ()
    provider: str | None = None
    source: str | None = None
    amenities: tuple[str, ...] = ()
    charging_power_kw: float | None = Field(default=None, ge=0)
    detour_minutes: float = Field(ge=0, default=0)
    distance_meters: float | None = Field(default=None, ge=0)
    charging_duration_minutes: float = Field(ge=0, default=0)
    mandatory: bool = False
    partner: "PartnerEnrichment | None" = None
    partner_benefit: str | None = None


class PartnerEnrichment(ContractModel):
    id: str
    name: str
    benefit: str | None = None
    benefit_scope: str | None = None
    benefit_source: Literal["fixture", "configured", "provider"] | None = None
    verified: bool = False
    status: Literal["suggested", "confirmed", "completed"] = "suggested"


class PartnerFact(ContractModel):
    partner_id: str
    brand: str
    benefit: str | None = None
    benefit_scope: str | None = None
    benefit_source: Literal["fixture", "configured", "provider"]
    verified: bool


class RouteOpportunity(ContractModel):
    id: str
    type: Literal["charging", "hotel", "restaurant", "amenity", "partner"]
    stop_id: str | None = None
    result_id: str | None = None
    partner_fact: PartnerFact | None = None
    reason: str = Field(min_length=1)
    detour_minutes: float = Field(ge=0, default=0)
    requires_route_confirmation: bool = False
    status: Literal["suggested", "confirmed", "completed"] = "suggested"


class ChargingPlan(ContractModel):
    stops: list[StopPinpoint] = Field(default_factory=list)
    complete: bool = False
    total_charging_minutes: float = Field(ge=0, default=0)
    confirmed: bool = False


class BorderCrossing(ContractModel):
    from_country: str
    to_country: str


class RouteRequirement(ContractModel):
    id: str
    name: str
    country: str
    kind: Literal["toll", "vignette"]
    mandatory: bool = True


class RouteSessionFacts(ContractModel):
    charging_plan_confirmed: bool = False
    confirmed_charging_stop_ids: list[str] = Field(default_factory=list)
    purchased_vignette_requirement_ids: list[str] = Field(default_factory=list)
    remaining_requirements: list[RouteRequirement] = Field(default_factory=list)


class TripStats(ContractModel):
    total_distance_km: float = Field(ge=0)
    driving_duration_minutes: float = Field(ge=0, default=0)
    total_duration_minutes: float = Field(ge=0)
    total_price_eur: float = Field(ge=0, default=0)


class TelemetryNarrationFacts(ContractModel):
    battery_percent: float = Field(ge=0, le=100)
    estimated_range_km: float = Field(ge=0)
    max_charged_range_km: float = Field(ge=0)
    consumption_rate_kwh: float = Field(gt=0)
    charging_feasible: bool | None = None


class RouteAlert(ContractModel):
    type: Literal["WEATHER", "TRAFFIC", "TOLL", "VEHICLE"]
    location_name: str | None = None
    severity: Literal["INFO", "WARNING", "CRITICAL"]
    message: str = Field(min_length=1)


class RouteResponse(ContractModel):
    origin: str = Field(min_length=1)
    destination: str = Field(min_length=1)
    stats: TripStats
    geometry: list[tuple[float, float]]
    stops: list[StopPinpoint] = Field(default_factory=list)
    alerts: list[RouteAlert] = Field(default_factory=list)
    border_crossings: list[BorderCrossing] = Field(default_factory=list)
    route_requirements: list[RouteRequirement] = Field(default_factory=list)
    charging_stop: StopPinpoint | None = None
    charging_required: bool = False
    opportunities: list[RouteOpportunity] = Field(default_factory=list)
    charging_plan: ChargingPlan | None = None
    session_facts: RouteSessionFacts | None = None
    telemetry: TelemetryNarrationFacts | None = None


class RealtimeToolRouteRequest(ContractModel):
    destination: str = Field(min_length=1, max_length=200)
    priority: RoutePriority = RoutePriority.BALANCED


class RealtimeToolRouteResponse(ContractModel):
    route: RouteResponse
    route_id: str | None = None
    search_id: str | None = None


class RealtimeToolSearchRoutePoiRequest(ContractModel):
    category: str = Field(min_length=1, max_length=50)
    location: str | None = None
    preference: str | None = None


class RealtimeToolSearchRoutePoiResponse(ContractModel):
    results: list[StopPinpoint] = Field(default_factory=list)
    opportunities: list[RouteOpportunity] = Field(default_factory=list)
    route_id: str | None = None
    search_id: str | None = None


class RealtimeToolSearchStopAmenitiesRequest(ContractModel):
    stop_id: str = Field(min_length=1, max_length=200)
    route_id: str = Field(min_length=1, max_length=200)
    search_id: str | None = Field(default=None, max_length=200)
    categories: list[str] = Field(default_factory=list, max_length=4)


class RealtimeToolSearchStopAmenitiesResponse(ContractModel):
    selected_stop_name: str = Field(min_length=1)
    results: list[StopPinpoint] = Field(default_factory=list, max_length=4)
    opportunities: list[RouteOpportunity] | None = None
    radius_meters: int = Field(default=500, ge=500, le=500)
    route_id: str
    search_id: str | None = None


class RealtimeToolConfirmChargingRequest(ContractModel):
    route_id: str = Field(min_length=1, max_length=200)
    stop_id: str | None = Field(default=None, max_length=200)
    confirmation: Literal["confirmed"]


class RealtimeToolConfirmChargingResponse(ContractModel):
    route: RouteResponse
    selected_stop_name: str
    results: list[StopPinpoint] = Field(default_factory=list, max_length=4)
    charging_plan: ChargingPlan | None = None
    session_facts: RouteSessionFacts | None = None
    radius_meters: int = Field(default=500, ge=500, le=500)
    route_id: str
    search_id: str | None = None


class RealtimeToolRerouteRequest(ContractModel):
    poi_id: str | list[str] = Field(min_length=1)
    route_id: str = Field(min_length=1, max_length=200)
    search_id: str = Field(min_length=1, max_length=200)
    coords: tuple[float, float] | None = None
    priority: RoutePriority | None = None
    confirmation: Literal["confirmed"]


class RealtimeToolRerouteResponse(ContractModel):
    route: RouteResponse
    route_id: str | None = None


class RealtimeToolPurchaseVignetteRequest(ContractModel):
    route_id: str = Field(min_length=1, max_length=200)
    requirement_id: str = Field(min_length=1, max_length=200)
    confirmation: Literal["confirmed"]


class RealtimeToolPurchaseVignetteResponse(ContractModel):
    status: Literal["completed", "duplicate"]
    transaction_id: str = Field(min_length=1)
    route_id: str
    requirement_id: str
    wallet_status: Literal["ready", "processing", "completed", "declined", "duplicate"]
    phone_confirmation_status: Literal["pending", "sent", "failed"]
    amount_eur: float = Field(ge=0)
    currency: Literal["EUR"]
    session_facts: RouteSessionFacts | None = None


class RealtimeToolBookingRequest(ContractModel):
    route_id: str = Field(min_length=1, max_length=200)
    search_id: str = Field(min_length=1, max_length=200)
    result_id: str = Field(min_length=1, max_length=200)
    booking_type: Literal["hotel_room", "restaurant_table"]
    guests: int = Field(ge=1, le=20)
    date: str = Field(min_length=1, max_length=30)
    time: str | None = Field(default=None, max_length=10)
    confirmation: Literal["confirmed"]


class RealtimeToolBookingResponse(ContractModel):
    status: Literal["completed", "duplicate"]
    booking_id: str = Field(min_length=1)
    result_id: str
    route_id: str
    booking_type: Literal["hotel_room", "restaurant_table"]
    guests: int = Field(ge=1, le=20)
    date: str
    time: str | None = None
    wallet_status: Literal["ready", "processing", "completed", "declined", "duplicate"]
    phone_confirmation_status: Literal["pending", "sent", "failed"]


class RealtimeToolStartDrivingRequest(ContractModel):
    route_id: str = Field(min_length=1, max_length=200)
    confirmation: Literal["confirmed"]


class DrivingNextStop(ContractModel):
    id: str
    name: str
    category: Literal[
        "charging", "hotel", "restaurant", "attraction", "coffee", "food",
        "rest", "toilets", "fuel", "toll", "vignette", "service", "shopping"
    ]


class RealtimeToolStartDrivingResponse(ContractModel):
    status: Literal["active"]
    route_id: str
    remaining_distance_km: float = Field(ge=0)
    remaining_duration_minutes: float = Field(ge=0)
    eta: str = Field(min_length=1)
    next_stop: DrivingNextStop | None = None
    charging_required: bool


class RealtimeToolReturnToMainRouteRequest(ContractModel):
    route_id: str = Field(min_length=1, max_length=200)


class RealtimeToolReturnToMainRouteResponse(ContractModel):
    status: Literal["success"]
    route_id: str


class RealtimeSessionResponse(ContractModel):
    client_secret: str = Field(min_length=1)
    model: str = Field(min_length=1)


class HealthResponse(ContractModel):
    status: str
    service: str
    environment: str


class ProviderHealthResponse(ContractModel):
    status: str
    environment: str
    openai_configured: bool
    google_routes_configured: bool
    google_places_configured: bool
    places_provider: str
    weather_provider: str = "offline"
    weather_configured: bool = False
    scenic_capability: bool = False
    partner_enrichment_ready: bool = False


class VehicleTelemetryResponse(ContractModel):
    vehicle_id: str
    propulsion: Literal["BEV"]
    battery_percent: float = Field(ge=0, le=100)
    estimated_range_km: float = Field(ge=0)
    max_charged_range_km: float = Field(ge=0)
    consumption_rate_kwh: float = Field(gt=0)
    tyres: Literal["SUMMER", "WINTER", "ALL_SEASON"]
    odometer_km: float = Field(ge=0)